### ISSUE TITLE
Have you considered joining the forces with similar GTK kotlin projects

### ISSUE THREAD

**Issue by leinardi** [vbsteven/gtk-kotlin-native#1]
Hi, I'm an Android developer and I would love to be able to write my GTK apps in Kotlin instead of Python. Looking around I saw there there are a couple of projects that are trying to make this happen but they are both a single developer projects in a very alpha stage.

Have you considered joining the forces with https://gitlab.com/gtk-kt/gtk-kt? I think that could make a huge difference to work on a singel project instead of implementing in parallel the same thing individually.

**Comment by vbsteven** [comment]
Hi @leinardi, thanks for suggesting this.

I am aware of the gtk-kt project and that project is mentioned in the readme. My original version was based on that repository but I decided to start from scratch for a few reasons: 

- As a learning opportunity about GTK. I had not done anything with GTK before this so trying to bridge GTK to KN forces me to learn the GTK API in depth.

- I wanted the freedom to experiment and refactor as I go. This project fits into a larger plan I have around Kotlin Multi Platform.

    Context: It is not open source yet but I'm working on a toolkit for cross-platform app development that targets web/desktop/mobile using Kotlin Multi Platform where KMP is used for all shared logic, I/O and platform-independent base components (forms, data store, list adapters, caching etc) combined with platform-specific implementations (GTK for desktop, Android Views for Android, UIKit for iOS, React/Vue/something for web). 

I am definitely open to collaboration with anyone interested but I'm going to keep the above points in mind before I decide to redirect my efforts to another project.

**Comment by leinardi** [comment]
Hi @vbsteven, thanks for the answer. I suggested that because writing all the bindings manually it's a huge work, and every time there is a new release of GTK, it will require again a huge effort. I think it would be way better to have a GIR parser that autogenerate the bindings. Such GIR Kotlin generator is something that your project could also take advantage, and making fully supporting new versions of GTK easier. What about cooperating with with the gtk-kt developer for his [gir-kt](https://gitlab.com/gtk-kt/gir-kt) project? That could bring you a big return of investment in effort you won't need to put in manually writing and, even more, maintaining for the GTK bindings...

**Comment by vbsteven** [comment]
I understand, and it is kind of where I want to go.

The goal should be to eventually have GIR-based autogenerated Kotlin bindings for GTK as you describe. But, in order to generate good bindings using GIR we need to know how the generated bindings API should work. And I'm not sure if I know that yet.

The easiest thing would be to simply wrap the existing C functions into Kotlin equivalents with lightweight wrapper classes (and possibly autogenerate them using GIR). This is what gtk-kt is doing and the current state of this project as well. This has some downsides:

- not idiomatic Kotlin
- lots of manual downcasting
- creating custom widgets and gobject subclasses is cumbersome
- not taking advantage of Kotlin features like Coroutines, Flow, and the type system

I think we can do better than that. So I'm trying to figure out a good idiomatic Kotlin API for GTK. One that takes advantage of modern language features:

- more type information in interfaces
    For example the GTK list store and model interface methods return object instances. I want to add generics to those.
- use Flow for input events
- a good solution for property binding
- coroutine and lifecycle support
    I would like to achieve something similar to AndroidX lifecycles and threading model where coroutine scopes are bound to the lifecycle of a window or widget.
- good support for subclassing widget and gobject
- maybe even TornadoFX-inspired typesafe CSS

You can view this project as step 1 in this process where I've done the initial binding work and have a library that can be used to write basic GTK applications.

My other project is step 2 where I am currently writing a real-world application using these bindings to see what works and what does not, changing the bindings as I go.

Step 3 is taking everything learned from 1 and 2 and use that as the target output for GIR-based bindings.

**Comment by leinardi** [comment]
Mmm I see your points, but I still think it would be good for you guys could have a chat about this topic. The current version of gtk-kt was all written manually by the developer. I think he also use that as a learning experience. But now he's working on the gir-kt project aiming to improve his previous work. So, don't judge the gtk-kt project form the actual published code because, once the gir-kt will be up and running, it will be very different.
Btw, I'm in touch with the dev because I'm helping with some small changes to the gir-kt (unfortunately I lack the C interop knowledge needed for a more meaningful contribution).
If you change your mind and want to get in touch, even just for a chat on the topic, let me know :slightly_smiling_face:.

**Comment by vbsteven** [comment]
Sure, I'm always up for a chat. You can reach me on the Kotlin Slack under my full name and my contact email is on my website which is linked in my Github bio.

**Comment by Doomsdayrs** [comment]
> more type information in interfaces
> For example the GTK list store and model interface methods return object instances. I want to add generics to those.

Can you elaborate on what you are thinking? Currently in gtk-kt lists are handled with respective wrapper classes that serve as extensions.

>use Flow for input events

As long as it is for input, using it for the activate function was a mistake I made long ago. In gtk-kt this can be done via a `callbackFlow` using the SignalManager return to close the signal.

> a good solution for property binding

Is what I have setup not good?

> coroutine and lifecycle support

I have started delving deeper into coroutines recently. Having successfully built a Dispatcher that will run coroutines on the main thread. To create a life cycle, would not be too hard. 

> good support for subclassing widget and gobject

What is good support?

> maybe even TornadoFX-inspired typesafe CSS

I agree that type safe CSS will be great in the future.

**Comment by leinardi** [comment]
Hi @vbsteven, I know that you prefer to wait a little before starting to mess with GIR, I just wanted to let you know that I started porting the very good GIR generator for java from [this](https://github.com/bailuk/java-gtk) project, to kotlin.

I made already some small progress: I managed to integrate KotlinPoet and [started to write a `CodeWriter` for it](https://github.com/leinardi/java-gtk/blob/kotlin-gtk/generator/src/main/kotlin/ch/bailu/gtk/writer/kotlin/KotlinApiWriter.kt).

I think it's coming along pretty decently, it can already generate all the files needed with proper package, KDoc and basic class declaration:
![image](https://user-images.githubusercontent.com/273338/219948847-9974e044-8f65-4bb3-9587-e88083046702.png)

But I lack the knowledge on how to deal with the C interop part required by the actual method implementation.

The `StructureModel` class provided by the GIR parser has already all the information required, I just need to figure out how to convert the JNA calls to Kotlin native one:
![image](https://user-images.githubusercontent.com/273338/219949040-0e2b7daa-f48f-4c80-b16c-1f64979ef0bb.png)

I was wondering if you could assist me with this task. Again, I think that using a GIR approach to generate the bindings could be a huge advantage for this project. I understand that you want to do with a more "custom tailored" approach for the bindings but those customizations could come as extension functions or, maybe, we can work together to merge them inside the code generation logic.

What do you think?

**Comment by vbsteven** [comment]
@leinardi do you have a specific repository/sample case/problem that I can look at or help with? 

You can probably get something up and running by combining the Kotlin/Native cinterop approach from this project (or gtk-kt) together with a GIR generator. 

This project uses the `cinterop` tool to generate KN versions of the GTK C methods under a specific package. For example the `adw_application_window_new(...)` C function from libadwaita gets a KN version called `adw_application_window_new(...)` in the `native.adwaita` package. (the generated package name is my own convention described in the readme and configured in the `adwaita.def` file.

So the generated `ApplicationWindow` constructor implementation from your last screenshot can acquire a `CPointer` to a new application window by calling `native.adwaita.adw_application_window_new(...)`.

I still need to take some time to organize my thoughts on all this. Simply generating the wrapper classes using GIR is only step 1.

Don't get me wrong, I really do want to have GIR for this project but I have not had the time to dive into it yet. Ideally someone comes up with a good GIR parser + basic Kotlin Poet setup so it can ingest a GIR library and output generated Kotlin files. Then I can take that and adapt the output Kotlin to the conventions in this project. (package names, method names, type conversions, object/widget subclassing, runtime type info for downcasting etc).

If you manage to get something like the above paragraph, I am **very** interested and we could probably collaborate/move pretty fast on it.

Extensions for additional functionality is what I was also thinking about. And some specialized classes like generic  versions for ListModel interfaces will need to be written by hand on top of the generated "untyped" interfaces. But all those can likely be added "on top" in an additional library that depends on the generated library.

Essentially it is a very layered approach: 

1. generate KN functions from a C library using `cinterop` (already done in this project for gtk/adwaita and some gobject/glib stuff) 
2. run GIR generator that outputs wrapper classes using KN functions (still TODO) 
3. compile GIR output + KN functions + extensions and utility classes into a final library. (this project already has all this aside from the GIR output from step 2.

So I think this project can get there soon with some help on the basic GIR setup. Let's say you manage to get a working bit of Kotlin code that parses a GIR library and passes it through a stubbed transform method, then I can probably take it from there to provide the actual generated implementation that conforms to the conventions of this project. After that I would be able to replace most of the manual wrapper classes with autogenerated ones and add the additional extensions+utillity classes on top.

That is the direction I would like to move in. What do you think?

**Comment by leinardi** [comment]
> @leinardi do you have a specific repository/sample case/problem that I can look at or help with?

Not yet but I should have one soon.

> You can probably get something up and running by combining the Kotlin/Native cinterop approach from this project (or gtk-kt) together with a GIR generator.

Yep, this is exactly what I'm doing :+1:.

> This project uses the cinterop tool to generate KN versions of the GTK C methods under a specific package. For example the adw_application_window_new(...) C function from libadwaita gets a KN version called adw_application_window_new(...) in the native.adwaita package. (the generated package name is my own convention described in the readme and configured in the adwaita.def file.
>
> So the generated ApplicationWindow constructor implementation from your last screenshot can acquire a CPointer to a new application window by calling native.adwaita.adw_application_window_new(...).

Yesterday, after some trial and error, I managed to successfully generate the KN versions of the C methods

> I still need to take some time to organize my thoughts on all this. Simply generating the wrapper classes using GIR is only step 1.
>
> Don't get me wrong, I really do want to have GIR for this project but I have not had the time to dive into it yet. Ideally someone comes up with a good GIR parser + basic Kotlin Poet setup so it can ingest a GIR library and output generated Kotlin files. Then I can take that and adapt the output Kotlin to the conventions in this project. (package names, method names, type conversions, object/widget subclassing, runtime type info for downcasting etc).

I'm currently working to get that done :slightly_smiling_face: 
My next milestone is to be able to get a simple HelloWord GTK4 screen to work with nothing on it (just a GTK Application and Window) from the generated code of the GIR parser. After that is done, it will be just a matter of expanding the code generation to add "[the rest of the owl](https://i.kym-cdn.com/photos/images/original/000/572/078/d6d.jpg)"

> If you manage to get something like the above paragraph, I am very interested and we could probably collaborate/move pretty fast on it.
>
> Extensions for additional functionality is what I was also thinking about. And some specialized classes like generic versions for ListModel interfaces will need to be written by hand on top of the generated "untyped" interfaces. But all those can likely be added "on top" in an additional library that depends on the generated library.
>
> Essentially it is a very layered approach:
>
> generate KN functions from a C library using cinterop (already done in this project for gtk/adwaita and some gobject/glib stuff)
> run GIR generator that outputs wrapper classes using KN functions (still TODO)
> compile GIR output + KN functions + extensions and utility classes into a final library. (this project already has all this aside from the GIR output from step 2.
> So I think this project can get there soon with some help on the basic GIR setup. Let's say you manage to get a working bit of Kotlin code that parses a GIR library and passes it through a stubbed transform method, then I can probably take it from there to provide the actual generated implementation that conforms to the conventions of this project. After that I would be able to replace most of the manual wrapper classes with autogenerated ones and add the additional extensions+utillity classes on top.
>
> That is the direction I would like to move in. What do you think?

Sounds like a plan! Do you have a preferred channel in case I get stuck and/or have some questions about C interop and Kotlin native? Last time we talked in Matrix but I'm not sure if you prefer to use something else (discord, email, this issue tracker, etc).

**Comment by vbsteven** [comment]
Alright! Let's draw this owl.

My matrix is a bit funky and I often miss messages. It is easier to reach me via email (address on my website in github bio) or on the Kotlin Slack.

**Comment by leinardi** [comment]
OK, I started drawing the first circles for our owl :) 

When you have time check out this project: https://gitlab.com/leinardi/ktgobject

I have committed [some generated bindings](https://gitlab.com/leinardi/ktgobject/-/blob/master/bindings/gtk/src/nativeMain/kotlin/binding/gtk/Application.kt). They are obviously broken but you can see that there has been some progress.

Right now the files are generated in the right directory, with the right name, package and, most of the times, constructor parameters. I still have to figure out how to handle some complex types like the one listed [here](https://gitlab.com/leinardi/ktgobject/-/blob/master/generator/src/main/kotlin/ch/bailu/gtk/table/WrapperTable.kt#L48).

Now I have to go but, if you have time, in the next days we can have a call and I can show you around. And maybe I can ask you a couple of questions about C interop topics.

**Comment by vbsteven** [comment]
Alright, I'll have a look later tonight.

Yesterday evening I hacked a little bit on it from scratch to understand the process a bit better.

Just plain read the GIR file as XML, parse it and generate files using Kotlinpoet. It does not seem that hard to do.

```
package io.quantus.gtk

import TODO.Window
import kotlinx.cinterop.CPointer

public class ApplicationWindow : Window {
  public val gtkApplicationWindowPointer: CPointer<*>
    get() = gPointer.asTypedPointer<GtkApplicationWindow>()

  public constructor(pointer: CPointer<*>) : super(pointer)
}
```

**Comment by leinardi** [comment]
the hardest part for me is to map all the C interop part, there are a bunch of types that I have no clue how should be mapped (see last link from my previous comment).

**Comment by vbsteven** [comment]
Yes, gobject/glib is the hard part. Gtk/Adwaita are "simple" since they are mostly GtkWidget subclasses with simple methods and properties/signals. These can be easily mapped onto Kotlin classes.

GObject is a full type system with all the necessary nuts and bolts to define new types (in C typically requiring lots of macros) and I don't think they need 100% mapping coverage for a usable Kotlin Native GTK library. A lot of these functions and types are never used by application developer code. This is also the part where some manual customization and utils could be useful.

I'm thinking maybe an iterative approach is in order here. Instead of trying to write a GIR generator with 100% coverage from the start, it might be better to focus on just the GTK and Adwaita GIR files (mostly widgets and some support types) and generate them from the (for now) manual Object base class and type registration that I've already built. Omitting methods with parameters/returntypes that the generator does not know about yet.

In essence trying to get the basic flow `GIR -> generated kotlin -> bindings library` compiling and working for minimal examples projects ASAP. And then iterate on the generator to add the missing types.

**Comment by leinardi** [comment]
> Omitting methods with parameters/returntypes that the generator does not know about yet.

The GIR generator is already doing that, and is writing comments on the top of the file with the types that are not handled:

![image](https://user-images.githubusercontent.com/273338/220304001-73e80b73-946d-4040-9163-f2018da889f8.png)


https://gitlab.com/leinardi/ktgobject/-/blob/master/bindings/glib/src/nativeMain/kotlin/binding/glib/ByteArray.kt#L2

**Comment by leinardi** [comment]
Hey, I made more progress with the generator. Now the constructors look way better and I fixed many parameters type.

I have a question for you: are you familiar with the concept of callback in this GTK / Kotlin native C interop?
Because I'm not, and I have no clue how they should be handler.

From what I saw, there are more or less 4 kind of classes that we should map from the GIR:

1. Record
2. Package
3. Callback
4. Pointer

Right now for Record and Pointer, I'm just holding an instance of a `CPointer` inside the class.

But for Callback I'm not sure what should be done. `java-gtk` has some complex logic to create callbacks that works with JNA. For now I dropped it and I'm just using the C interop types directly (e.g. `glib.GDestroyNotify`, needed by `gtk_tree_list_model_new()`) but I'm not sure if this is a good idea.

Do you have any experience with this kind of callbacks?

**Comment by vbsteven** [comment]
sidenote: I have some urgent client work I need to finish by next week so the next few days I don't have much time to hack on it myself. I hope to have another look at the GIR generator next week.

Yes, I am familiar with the GTK callback mechanism and C interop. In this repo you can find examples on how to use them from KN. 

Have a look at the [GTK Button](https://github.com/vbsteven/gtk-kotlin-native/blob/master/gtk-bindings/src/nativeMain/kotlin/bindings/gtk/Button.kt) implementation which uses a `GCallback` to connect a Kotlin lambda to to the button `clicked` signal. Mainly the `onClicked` method and the associated `staticButtonClickedFunc` in the same file and `staticStableRefDestroy` in Signals.kt.

```kotlin
    /**
     * Connect [func] as a handler for the "clicked" signal.
     */
    fun onClicked(func: (Button) -> Unit) {
        g_signal_connect_data(
            gPointer,
            "clicked", 
           staticButtonClickedFunc,
            StableRef.create(func).asCPointer(),
            staticStableRefDestroy,
            0
        )
    }
```
In essence, the `g_signal_connect_data` C function takes: 

1. a pointer to the object you want to attach a signal to (in this case the pointer for the Button instance)
2. the name of the signal,
3. a pointer to a C function that conforms to the `GCallback` C type (the C function that will be called by GTK when the signal is emitted)
4. a pointer to userdata that GTK will pass as an argument when invoking the callback function from 3
5. a pointer to a C function to clean up any memory when the callback isn't needed anymore (conforming to `GClosureNotify`)
6. flags for the signal connect

`StableRef` is important here. StableRef is a Kotlin C interop mechanism to create a Stable reference to a Kotlin object so you can safely pass a C pointer pointing to that object to a C function without it being garbage collected. (Don't forget to `dispose()` the StableRef afterwards or you'll leak memory, this is what I do inside `staticStableRefDestroy`).

`staticCFunction` is another important Kotlin C interop feature that creates C functions that you can pass to other C functions as a function pointer.

```kotlin
private val staticButtonClickedFunc: GCallback =
    staticCFunction { buttonPointer: CPointer<GtkButton>?,
                      data: gpointer? ->
        val button = buttonPointer!!.asButton()
        val func = data!!.asStableRef<(Button) -> Unit>().get()
        func.invoke(button)
    }.reinterpret()
```

This bit of code creates a C function conforming to `GCallback` that can be passed to other C functions as a function pointer. When the lambda is invoked, it extracts the original user-provided Kotlin callback, and tries to invoke it.

**Comment by vbsteven** [comment]
To clarify: I don't think a GIR generator has to create wrapper classes for `GCallback` and and other callback types. Instead the necessary glue code needs to be generated for object methods that take a callback as an argument to wrap the userprovided Kotlin lambda into a C function + userdata pointer combo and handle the required data marshaling and reference cleanup.

This part is one of the reasons that I mentioned earlier why I'm currently still experimenting with manual bindings to figure out the best mappings from C to Kotlin before I fully commit to GIR.

For example: 

Should every signal be mapped to a method on the wrapper class like I did with `Button.onClicked`? 
Should it be called `onClicked` or something else? 
For a click handler like this example, does it have to be a no arguments lambda or do we pass in the `Button` instance as first argument? (maybe both?)
How should it behave when the user provided lambda throws an Exception and the C callback requires a Boolean (or other type) return value.

These types of questions are still unclear to me.

Button `clicked` is the easiest version btw, the C version is a simple callback with only the Button pointer as first argument. But other signals can be more complex with additional arguments and return types.

**Comment by leinardi** [comment]
Hey thanks a lot! This info is very useful! And don't worry about the urgent work, I'll be also pretty busy for the rest of the week. We will draw this own :smile:.

**Comment by leinardi** [comment]
BTW one more question: do you know how to deal with Classes that don't have a native construction? There are few of them but I have currently no clue on how to deal with them:
![image](https://user-images.githubusercontent.com/273338/220914040-2a169204-e391-484d-9a1c-a5cc74074cc7.png)

`java-gtk` is doing some black magic (creating some JNA allocators):
![image](https://user-images.githubusercontent.com/273338/220914871-9e96f46e-9719-419f-b399-ae169f3bf7b5.png)

But maybe there is an easier way with Kotlin native...

**Comment by vbsteven** [comment]
Depends on the use case.

`NativeDialog` is an abstract class which cannot be instantiated by itself. Its wrapper class only needs a constructor taking a `CPointer<GtkNativeDialog>` so it can be called as `super(pointer)` from the constructors of its descendant classes and so you can convert a pointer to an instance of the Wrapper class.

`TextIter` is a C struct. Again one of these things I'm not 100% sure about how to properly do that yet to cover most cases. Should these structs be mapped to wrapper Kotlin classes or not. (same for enums en flags)

Working with C structs in Kotlin Native requires some cinterop code.

```kotlin
    fun getIntProperty(name: String): Int = memScoped {
        val gValue = alloc<GValue>()
        g_value_init(gValue.ptr, G_TYPE_INT)
        g_object_get_property(gPointer.asTypedPointer(), name, gValue.ptr)
        return g_value_get_int(gValue.ptr)
    }
```

This is an example from `Object.kt` in this repo that works with a GValue struct. 

It uses `memScoped {}` to create a scope in which you can allocate memory on the C side and work with the struct.
It first allocates memory for the struct, and then calls `g_value_init` to initialize it (GTK specific thing), and then it uses `gValue.ptr` to get a CPointer to that struct that can be passed to other C functions.

Another example that I have not committed to master yet is `adw_swipeable_get_swipe_area` from the Adwaita Swipeable interface. It is a method that requires an out pointer to a `GdkRectangle` struct that the function will populate with its result. I handled this case by creating a Kotlin data class representing a Rectangle and abstracted the struct allocation, pointer passing and extraction to data class in the method implementation so the user does not need to care about this struct stuff and can just call a method and get the Rectangle back.

```kotlin
/**
 * Data class equivalent to GdkRectangle struct
 */
data class Rectangle(
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
)

// in the Swipeable interface

fun getSwipeArea(direction: AdwNavigationDirection, isDrag: Boolean): Rectangle = memScoped {
    val outRect = alloc<GdkRectangle>()
    adw_swipeable_get_swipe_area(adwSwipeablePointer, direction, isDrag.gboolean, outRect.ptr.reinterpret())
    return Rectangle(outRect.x, outRect.y, outRect.width, outRect.height)
}
```

**Comment by leinardi** [comment]
Thanks! Again, this is incredibly useful to me since I lack a lot of knowledge of C interop!
I'm now taking care of the methods and functions generations but I'll come back to this soon!

**Comment by EsoCoding** [comment]
Just a tip. The D language has [girtod](https://github.com/gtkd-developers/gir-to-d/blob/master/source/girtod.d) that generates bindings for gtk4, i would suggest looking at its code, its readable code even if your not into D language and the code is not to complex. Even though i'm no coding expert i understand it for a large part. I'm interested in learning kotlin, but missing bindings is the reason it's still a no go. Gtk is to important for me. The D language has reputation problem and misses popularity, and many projects seem to be dead. Also GtkD has had no commits for last year. Yet it has working gir generated gtk4 bindings which many languages still not have finished. So might be worth checking it out.

**Comment by vbsteven** [comment]
Thanks for the tip @EsoCoding, I will definitely have a look at how the D language did it for inspiration.

I don't have an abundance of free time lately but managed to get a good hacking session in this morning and it is progressing well. I think I can have a minimal working generator up and running soon that should be able to replace the manual bindings I already have in this repository.

**Comment by leinardi** [comment]
Hey @vbsteven, I'm getting closer to have some buildable code for the GLib generated classes but I'm now facing an issue with native calls to functions that have a `vararg` as parameter.

For example `gtk_dialog_add_buttons`, which is mapped with this:
```kotlin
@kotlinx.cinterop.internal.CCall public external fun gtk_dialog_add_buttons(dialog: kotlinx.cinterop.CValuesRef<native.gtk.GtkDialog /* = native.gtk._GtkDialog */>?, @kotlinx.cinterop.internal.CCall.CString first_button_text: kotlin.String?, vararg variadicArguments: kotlin.Any?): kotlin.Unit { /* compiled code */ }
```

It seems to expect a `vararg variadicArguments: kotlin.Any?`, but if I send that to it like this:
```kotlin
    public fun addButtons(firstButtonText: String, vararg variadicArguments: Any): Unit {
        gtk_dialog_add_buttons(cPointer.reinterpret(), firstButtonText, variadicArguments)
    }
```
I get this error:
```
type kotlin.Array<out kotlin.Any>  is not supported here: doesn't correspond to any C type
```

The GIR parameter only gives a name (`...`) and has a tag `varargs`:
```xml
          <parameter name="..." transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="../gtk/gtkdialog.c"
                 line="950">response ID for first button, then more text-response_id pairs</doc>
            <varargs/>
          </parameter>
```

Do you have any clue of what should be sent there?

**Comment by vbsteven** [comment]
I think the vararg parameter is always the last one and should be applied to the parameter before it.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by leinardi** [comment]
Unfortunately isn't an easy solution, but I think I found a workaround. I'll tell you more this evening :)

**Comment by vbsteven** [comment]
I'm closing this now that we have opened up the gtk-kn repository. 

https://gitlab.com/gtk-kn/gtk-kn
